docs: add guide to setup windows machine for WSL (#2876)
canonical/ubuntu-pro-client: dev-docs/howtoguides/how_to_setup_windows_machine_for_wsl_testing.md

> # How to setup Windows machine for WSL testing

Reviewer: ```suggestion
# How to set up a Windows machine for WSL testing
```

To run our integration tests on WSL instances, we need to first setup a Windows machine
to allows us to launch the WSL instances. Not only that, we also need to setup OpenSSH
server to allow us to run ssh commands on the Windows machine.

> This how to guide will explain all the necessary steps for the setup, from launching the image
> from Azure to testing if WSL is correctly working on the machine.

Reviewer 2: ```suggestion
This how-to guide will explain all the necessary steps for the setup, from launching the image
in Azure to testing if WSL is correctly working on the machine.
```

Reviewer 3: I guess this is a matter of taste, but wouldn' tit read better if we drop the `how-to` part of the sentence : `This how-to guide will explain all the necessary step...`. Also not being an explanation, but rather a how-to guide, shouldn't we rephrase that sentence into `This guide presents all the necessary steps...` ?

Author: Yes, it is the latter. We are launching a Windows machine in Azure. And I will also specify the Windows version here as well

## Launching the image on Azure

When launching the Windows instance on Azure, there a few configurations that need to be performed,
otherwise we will not be able to launch WSL instances.

> First of all, name the machine as `wsl-test`. Currently, this is the name we expect on the

Author: I don't think there is a specif link to launch a Windows machine, but this is one the first steps when launching a image in Azure, that's why I have not added any screenshots here. But I don't have a strong opinion. If you believe it is worth, I can add it. No problem at all

> First of all, name the machine as `wsl-test`. Currently, this is the name we expect on the

Reviewer: In that case, I think it could be helpful to link to a guide on launching images in Azure if you know of a good one. It would be ++convenient for a reader who may not have used Azure before if we can cut out a bit of Googling for them, but it's not essential.

integration tests.

After that, select the `Windows 11` image and mark the `Security Type` as *Standard*.

> ![windows_secuirty_type](windows-security-type.png)

Reviewer 2: ```suggestion
![windows_security_type](windows-security-type.png)
```

Reviewer 3: I'd add the following to this step:

`The alternative "Trusted Launch" [doesn't support Nested Virtualization](https://learn.microsoft.com/en-us/azure/virtual-machines/trusted-launch#unsupported-features), required to run WSL.`

or something similar.

Although this is not an explanation documentation, if the user skips this step it can be really hard to figure out what is causing the error when trying to set up a WSL instance.

Author: Not really, I will leave it blank

> Set an username and password for the VM. Please set the username as *ubuntu*. 

Reviewer 2: ```suggestion
Set a username and password for the VM. Please set the username as *ubuntu*. 
```

We will use this account setting to manually login into the machine to run some commands there directly.

> ![windows_secuirty_type](windows-ssh-setup.png)

Reviewer 2: ```suggestion
![windows_ssh_setup](windows-ssh-setup.png)
```

Author: I am leaving this blank as @s-makin suggested

Finally, remember to also select the SSH inbound port for the machine

> ![windows_secuirty_type](windows-setup-port.png)

Reviewer 2: ```suggestion
![windows_setup_port](windows-setup-port.png)
```

> ## Configure Bastion to login into the machine

Reviewer: ```suggestion
## Configure Bastion to log into the machine
```

> ## Configure Bastion to login into the machine
> 
> To login into the machine and access its GUI, you need to setup Bastion.

Reviewer 2: ```suggestion
## Configure Bastion to log into the machine

To log into the machine and access its GUI, you need to setup a Bastion.
```

> After launching the machine, go for the *Connect with Bastion* option:

Reviewer: ```suggestion
After launching the machine, go for the *Connect via Bastion* option:
```

> ![windows_secuirty_type](windows-bastion.png)
> ![windows_secuirty_type](windows-deploy-bastion.png)

Reviewer 2: ```suggestion
![windows_bastion](windows-bastion.png)
![windows_deploy_bastion](windows-deploy-bastion.png)
```

Reviewer: This renders a bit weirdly. I'd add empty lines between 32-33 and 33-34.

> After deploying Bastion, you just need to present your username and password to login into the

Reviewer 2: ```suggestion
After deploying the Bastion, you just need to present your username and password to log into the
```

machine.

## Configuring Hyper-V

> After login into the machine through Bastion, you need to configure Hyper-V by following

Reviewer 2: ```suggestion
After logging into the machine through Bastion, you need to configure Hyper-V by following
```

> this [tutorial](https://learn.microsoft.com/en-us/azure/lab-services/how-to-enable-nested-virtualization-template-vm-using-script?tabs=powershell)

Reviewer 3: I'm missing a reference to the `Virtual Machine Platform` optional feature. Does this step cover it?

We had this error when setting up a VM, fixed by setting that feature ON, if you remember:

```
WslRegisterDistribution failed with error: 0x80070424
Error: 0x80070424 The specified service does not exist as an installed service.
```

Author: What I have found is that when just running `wsl --install`, it already covers this step automatically.
But I will let @orndorffgrant and @renanrodrigo try this out first just to guarantee if this is indeed correct

Reviewer 4: According to the Microsoft Docs, `--install` does it for you. https://learn.microsoft.com/en-us/windows/wsl/install#install-wsl-command.

@CarlosNihelton and I are just not used to these new conveniences 😅.

Reviewer 3: Indeed 😄

## Installing winget

> For each Ubuntu distro, we will get the installers directly from the Microsoft store.

Reviewer: I'm not 100% sure I understand what you mean by this. Do you mean each Ubuntu release has an installer in the Microsoft store?

Author: Yes, exactly. I will make this clearer in the guide

To install such installers, we need to configure *winget*.

To install *winget* on the machine, follow this [tutorial](https://winget.pro/winget-install-powershell/)

```{note}
if you see the following error when installing *winget*, restart the machine first:

Add-AppxPackage : Deployment failed with HRESULT: 0x80073D02, The package could not be installed
because resources it   modifies are currently in use. 

> ```

Reviewer: I recommend putting this into a code block. When we uprate the dev docs to actual docs, this will throw up spelling errors if we don't.

> ## Installing wsl

Reviewer: ```suggestion
## Installing WSL
```

> To install *wsl*, just run the following command, as an administrator, on powershell:

Reviewer: ```suggestion
To install WSL, just run the following command, as an administrator, on powershell:
```

```console
$ wsl --install
```

> After that, we need to update WSL to the pre-release version by running:

Reviewer: Why do we want the pre-release version?

Author: I am adding an explanation for why this is necessary

```console
$ wsl --update --pre-release
```

Now, install a Bionic instance to test the whole WSL setup is working as expected:

```console

> $ winget install --id "9PNKSF5ZN4SW" --accept-source-agreements --accept-package-agreements --silent

Reviewer 2: I'm guessing this ID refers to the bionic wsl app? I think that deserves a short explanation and maybe also an explanation of how you got this ID

Reviewer 4: Note that `winget install --name "Ubuntu 18.04 LTS" --exact ...` would also work.

Author: That's a great suggestion for this guide. I will use it. Thanks @EduardGomezEscandell

```

> Now, create the image with:
> 
> ```console
> $ & ubuntu1804.exe install --root --ui=none

Reviewer 3: ```suggestion
Now, create the instance with:

```console
$ & ubuntu1804.exe install --root
```

When using `install --root` the `--ui` option is disregarded. Also we plan to remove the UI anyways, so this will make your documentation future-proof.

```

After that, just test to see if we can run commands on the WSL instance:

```console

> $ wsl -d Ubuntu-18.04 --exec lsb_release -a

Reviewer 3: Note that running `ubuntu1804.exe run` instead of `wsl -d Ubuntu-18.04 --exec` is equally valid.

Author: That is totally fair. I am using `wsl` here just because this is what I am doing on the integration tests

```

## Installing OpenSSH server

To install OpenSSH server, follow these steps:

1) Open *Settings*
2) Go to *Apps*
3) Go to *Optional Features*
4) Go to *Add an optional feature*
5) Search for *ssh*
6) Install OpenSSH Server

> After installing the Service, open PowerShell and start-it by running:

Reviewer: ```suggestion
After installing the Service, open PowerShell and launch it by running:
```

```console
$ Start-Service sshd
```

And set it to automatically start after boot:

```console
$ Set-Service -Name sshd -StartupType 'Automatic'
```

### Configure the SSH keys

We need to add a SSH public key into the Windows machine.

To do that, follow these steps:

> 1) Send the public key to the Windows machine:

Reviewer 2: > the public key

which public key? should I create one somewhere first?

Author: Yep, you should create one or reuse an existing key. I will this explicit in the guide

```console

> $ scp PUB_KEY_PATH ubuntu@IP:%programdata%/ssh

Reviewer 2: ```suggestion
$ scp PUB_KEY_PATH ubuntu@$WINDOWS_MACHINE_IP_ADDRESS:%programdata%/ssh
```
to make it more obvious that IP needs to be replaced

Reviewer 2: or maybe explain the  `UACLIENT_BEHAVE_WSL_IP_ADDRESS` environment variable up here and use that

Reviewer: I'd suggest both

```

> 2) SSH into the Windows machine

Reviewer 2: Is this suggesting SSH-ing with username and password since the key isn't set up yet?

Author: Yes, it is. I will make that explicit as well

3) Run the following commands:

```console
$ cd %programdata%/ssh
```

```console
$ type PUB_KEY_NAME >> administrators_authorized_keys
```

```console
$ icacls administrators_authorized_keys /inheritance:r /grant "Administrators:F" /grant "SYSTEM:F"
```

Now test SSH into the machine using your private key

> ## Configure the machine to automatic logon

Reviewer: ```suggestion
## Configure the machine for automatic login
```

> To properly run *winget* commands after a machine reboot, an user

Reviewer 4: ```suggestion
To properly run *winget* commands after a machine reboot, a user
```

needs to be already logged into the machine. To automate that process,
follow this [tutorial](https://learn.microsoft.com/en-us/troubleshoot/windows-server/user-profiles-and-logon/turn-on-automatic-logon#use-registry-editor-to-turn-on-automatic-logon)

> Note that you also need to manually create the *DefaultUserName* and *DefaultPassword* files

Reviewer: ```suggestion
Note that you also need to manually create the *DefaultUserName* and *DefaultPassword* files.
```

## Running the integration test

After this setup, you should be able to use that machine to run the WSL tests automatically.

> Before running a WSL test, remember to set the following environment variables:

Reviewer 3: Set the environment variables inside each WSL instance subject to the integration tests or in the Windows host environment as well?

I suspect it's only the first, but wouldn't it be better to clarify it?

Author: It is actually neither of the options, we need to set up that up on your local machine. Our integration tests will use them to start and communicate with the Windows host machine and launch the WSL tests.

But I will make this more explicit

# How to set up a Windows machine for WSL testing

To run our integration tests on Windows Subsystem for Linux (WSL) instances, we need to first set up
a Windows machine that will allow us to launch the WSL instances. After that, we also need to set up
an OpenSSH server so we can run SSH commands on the Windows machine.

This guide presents all the necessary steps for the set up, from launching the Windows 11 
image in Azure to testing if WSL is correctly working on the machine.

## Launching the image on Azure

When launching the Windows instance on Azure, there are a few configurations that need to be performed,
otherwise we will not be able to launch WSL instances.

First of all, name the machine as `wsl-test`. Currently, this is the name we expect on the
integration tests.

After that, select the `Windows 11` image and mark the `Security Type` as *Standard*.

> ![](windows-security-type.png)

Reviewer 2: The screenshow shows Ubuntu not windows.
Also I see several `Windows 11`s
* Windows 11 Pro
* Windows 11 Pro N
* Windows 11 Enterprise
* Windows 11 Enterprise multi-session
* Windows 11 Enterprise N

I'll try out the Windows 11 Pro for now but it would be good to choose a specific one for the guide

We will use this account setting to manually log into the machine to run some commands there directly.
![](windows-ssh-setup.png)

Finally, remember to also select the SSH inbound port for the machine
![](windows-setup-port.png)

## Configure Bastion to log into the machine

To log into the machine and access its GUI, you need to set up Bastion.
After launching the machine, go for the *Connect via Bastion* option:

![](windows-bastion.png)

![](windows-deploy-bastion.png)

After deploying the Bastion, you just need to present your username and password to log into the
machine.

## Configuring Hyper-V

After logging into the machine through Bastion, you need to configure Hyper-V by following
this [tutorial](https://learn.microsoft.com/en-us/azure/lab-services/how-to-enable-nested-virtualization-template-vm-using-script?tabs=powershell)

## Installing winget

Each WSL Ubuntu distro has a dedicated installer in the Microsoft store. Therefore, before launching
any WSL instance, we need to install the installers first. To achieve that, we need to configure
**winget**.

To install it on the machine, follow this [tutorial](https://winget.pro/winget-install-powershell/)

> ```

Reviewer: ```suggestion
```{note}
```

> if you see the following error during installation, restart the machine first:

Reviewer: ```suggestion
If you see the following error during installation, restart the machine first:
```

Add-AppxPackage : Deployment failed with HRESULT: 0x80073D02, The package could not be installed
because resources it modifies are currently in use. 
```

## Installing WSL

To install WSL, just run the following command, as an administrator, on powershell:

```console
$ wsl --install
```

After that, we need to update WSL to the pre-release version by running:

```console
$ wsl --update --pre-release
```

```{note}

> We are installing the pre-release version because it allow us to directly run wsl commands

Reviewer: ```suggestion
We are installing the pre-release version because it allow us to directly run WSL commands
```

through SSH. Once that fix is officially release, we will no longer need to install the
pre-release version.
```

Now, install a Bionic instance to test the whole WSL set up is working as expected:

```console
$ winget install --name "Ubuntu 18.04 LTS" --accept-source-agreements --accept-package-agreements --silent
```

Now, create the instance with:

```console
$ & ubuntu1804.exe install --root
```

After that, just test to see if we can run commands on the WSL instance:

```console
$ wsl -d Ubuntu-18.04 --exec lsb_release -a
```

## Installing OpenSSH server

To install OpenSSH server, follow these steps:

> 1) Open *Settings*

Reviewer: I think I might have overlooked something, but what application are we using here?

2) Go to *Apps*
3) Go to *Optional Features*
4) Go to *Add an optional feature*
5) Search for *ssh*
6) Install OpenSSH Server

After installing the Service, open PowerShell and launch it by running:

```console
$ Start-Service sshd
```

And set it to automatically start after boot:

```console
$ Set-Service -Name sshd -StartupType 'Automatic'
```

### Configure the SSH keys

We need to add a SSH public key into the Windows machine.

To do that, follow these steps:

> 1) Create a public key to be used in the Windows machine (You can also reuse an existing key if you want to)

Reviewer: ```suggestion
1) Create a public key to be used in the Windows machine (You can also reuse an existing key if you want to).
```

2) Set the **UACLIENT_BEHAVE_WSL_IP_ADDRESS** environment variable with the Windows machine IP address. 
This variable will be used in the integration test, but we will already use it in one of the
following steps.

2) Send the public key to the Windows machine:

```console
$ scp PUB_KEY_PATH ubuntu@$UACLIENT_BEHAVE_WSL_IP_ADDRESS:%programdata%/ssh
```

3) SSH into the Windows machine. Note that this SSH will ask you for the password you have set up during the Windows image creation.
4) Run the following commands:

```console
$ cd %programdata%/ssh
```

```console
$ type PUB_KEY_NAME >> administrators_authorized_keys
```

```console
$ icacls administrators_authorized_keys /inheritance:r /grant "Administrators:F" /grant "SYSTEM:F"
```

> Now test SSH into the machine using your private key

Reviewer: ```suggestion
Now test SSH into the machine using your private key.
```

## Configure the machine for automatic login

To properly run **winget** commands after a machine reboot, a user
needs to be already logged into the machine. To automate that process,
follow this [tutorial](https://learn.microsoft.com/en-us/troubleshoot/windows-server/user-profiles-and-logon/turn-on-automatic-logon#use-registry-editor-to-turn-on-automatic-logon)

Note that you also need to manually create the **DefaultUserName** and **DefaultPassword** files.

## Running the integration test

After this set up, you should be able to use that machine to run the WSL tests automatically.
Before running a WSL test, remember to set the following environment variables:

* **UACLIENT_BEHAVE_WSL_IP_ADDRESS**
* **UACLIENT_BEHAVE_WSL_PRIVKEY_PATH**
* **UACLIENT_BEHAVE_WSL_PUBKEY_PATH**

> Note that this variables must be set on our local machine. The integration test setup will use them

Reviewer: ```suggestion
Note that these variables must be set on your local machine. The integration test setup will use them
```

> to stablish the connection with the Windows host machine to create the WSL instances the tests will

Reviewer: ```suggestion
to establish the connection with the Windows host machine, to then create the WSL instances the tests will
```

Add-AppxPackage : Deployment failed with HRESULT: 0x80073D02, The package could not be installed
because resources it modifies are currently in use. 
```

## Installing WSL

To install WSL, just run the following command, as an administrator, on powershell:

```console
$ wsl --install
```

After that, we need to update WSL to the pre-release version by running:

```console
$ wsl --update --pre-release
```

```{note}
We are installing the pre-release version because it allow us to directly run wsl commands
through SSH. Once that fix is officially release, we will no longer need to install the
pre-release version.
```

Now, install a Bionic instance to test the whole WSL set up is working as expected:

```console
$ winget install --name "Ubuntu 18.04 LTS" --accept-source-agreements --accept-package-agreements --silent
```

Now, create the instance with:

```console
$ & ubuntu1804.exe install --root
```

After that, just test to see if we can run commands on the WSL instance:

```console
$ wsl -d Ubuntu-18.04 --exec lsb_release -a
```

## Installing OpenSSH server

To install OpenSSH server, follow these steps:

1) Open *Settings*

> 2) Go to *Apps*

Reviewer 2: For me, Optional Features is in the "System" category, not the "Apps" category

Reviewer 5: This is correct - I found it in *Sysytem*

3) Go to *Optional Features*
4) Go to *Add an optional feature*
5) Search for *ssh*
6) Install OpenSSH Server

After installing the Service, open PowerShell and launch it by running:

```console
$ Start-Service sshd
```

And set it to automatically start after boot:

```console
$ Set-Service -Name sshd -StartupType 'Automatic'
```

### Configure the SSH keys

We need to add a SSH public key into the Windows machine.

To do that, follow these steps:

1) Create a public key to be used in the Windows machine (You can also reuse an existing key if you want to)

2) Set the **UACLIENT_BEHAVE_WSL_IP_ADDRESS** environment variable with the Windows machine IP address. 
This variable will be used in the integration test, but we will already use it in one of the
following steps.

2) Send the public key to the Windows machine:

```console

> $ scp PUB_KEY_PATH ubuntu@$UACLIENT_BEHAVE_WSL_IP_ADDRESS:%programdata%/ssh

Reviewer 2: This didn't work for me but the following did
```
scp .ssh/wsl_test_rsa.pub ubuntu@$UACLIENT_BEHAVE_WSL_IP_ADDRESS:C:\\ProgramData\\ssh
```

Reviewer 5: both work for me

Author: I am changing the option that works for everyone
